hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/multipart/MultipartDownloaderTest.java

```diff
@@ -0,0 +1,130 @@
+package org.apache.hadoop.fs.s3a.multipart;
+
+import com.google.common.base.Charsets;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.Assert.assertTrue;
+
+public class MultipartDownloaderTest {
+
+    private final byte[] bytes = generateBytes(0, 1000);
+
+    private final MultipartDownloader multipartDownloader = new MultipartDownloader(
+            100,
+            MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(2)),
+            new PartDownloader() {
+                @Override
+                public InputStream downloadPart(String bucket, String key, long rangeStart, long rangeEnd) {
+                    return new ByteArrayInputStream(Arrays.copyOfRange(bytes, (int) rangeStart, (int) rangeEnd));
+                }
+            },
+            10,
+            1000);
+
+    @Test
+    public void testDownloadAll() throws IOException {
+        assertDownloadRange(0, bytes.length);
+    }
+
+    @Test
+    public void testDownloadRange() throws IOException {
+        assertDownloadRange(10, 20);
+    }
+
+    @Test
+    public void testDownloadTail() throws IOException {
+        assertDownloadRange(bytes.length - 10, bytes.length);
+    }
+
+    @Test
+    public void testCancelPartsIfFailure() throws InterruptedException {
+        PartDownloader partDownloader = new PartDownloader() {
+            @Override
+            public InputStream downloadPart(String bucket, String key, long rangeStart, long rangeEnd) {
+                if (rangeStart == 0) {
+                    throw new RuntimeException();
+                } else {
+                    return new InputStream() {
+                        @Override
+                        public int read() {
+                            try {
+                                Thread.sleep(1000000000);
```

Comment:
this tips the test into being a subclass of `S3AScaleTestBase` for the longer, configurable test timeouts, and only running if -Dscale is set (i.e. less often). Better: come up with a way to implement delays somehow

Reply:
The sleep is really there just to block forever. The main test will only wait 1 second to verify that the interrupt was delivered. If it is working, there is minimal wait (not even 1 second), if it fails, it will only wait 1 second. I changed the Thread.sleep() to a proper indefinite block to make it clear that we don't expect it.